### PR TITLE
Support 4 players in mame2010

### PR DESCRIPTION
**Pull request by markwkidd**
Closes https://github.com/libretro/mame2010-libretro/issues/37
Closes https://github.com/libretro/mame2010-libretro/issues/21

@r-type I had time to test this with a joypad today. From my testing it seems that this approach will allow not just the WiiU can use the MAME UI select but for any joypad.

For Joypad 1 , the RetroPad A doubles as the `IPT_UI_SELECT`

(Edit: Whoa -- I didn't realize that submitting this PR would 'suck up' my prior conversation with @r-type. Well, here is is, the whole record of our convo.)

### CODE REVIEW

**Comment by markwkidd** [comment]
@r-type I had time to test this with a joypad today. From my testing it seems that this approach will allow not just the WiiU can use the MAME UI select but for any joypad.

For Joypad 1 , the RetroPad A doubles as the IPT_UI_SELECT

File: src/osd/retro/retromain.c [hunk 1]
```diff
@@ -1562,11 +1513,45 @@ void retro_poll_mame_input()
    P2_state[KEY_BUTTON_4]   = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_Y);
    P2_state[KEY_BUTTON_5]   = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_L);
    P2_state[KEY_BUTTON_6]   = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_R);
+   P2_state[KEY_BUTTON_7]   = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_L2);
    P2_state[KEY_JOYSTICK_U] = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_UP);
    P2_state[KEY_JOYSTICK_D] = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_DOWN);
    P2_state[KEY_JOYSTICK_L] = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_LEFT);
    P2_state[KEY_JOYSTICK_R] = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_RIGHT);
 
+   P3_state[KEY_TAB]        = input_state_cb(1, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_L2);
```
[the remark this answers is not in the document]
Comment (markwkidd) [reply src/osd/retro/retromain.c]: Ah that is a good catch! Fixed via: https://github.com/libretro/mame2010-libretro/pull/102/commits/855ad48450dc44994c1d34220a632f60550ec77d

Somehow even with this error I was able to add coins from P3 so I hand't caught the error yet. BTW I am trying to find a good 4+ player game to use as a reference test for this PR. I'll post about that in the main discussion.

File: src/osd/retro/retromain.c [hunk 2]
```diff
@@ -1576,25 +1561,41 @@ void retro_poll_mame_input()
...
          if (PLAYER2_PRESS(B))
             P2_state[KEY_BUTTON_2] = turbo_state < turbo_delay ? 0 : 1;
-         break;
+         if (PLAYER3_PRESS(B))
+            P3_state[KEY_BUTTON_3] = turbo_state < turbo_delay ? 0 : 1;
+         if (PLAYER4_PRESS(B))
+            P4_state[KEY_BUTTON_2] = turbo_state < turbo_delay ? 0 : 1;
+        break;
       case 3:
          if (PLAYER1_PRESS(R2))
```
[the remark this answers is not in the document]
Comment (markwkidd) [reply src/osd/retro/retromain.c]: Thank you for calling my attention to this -- I originally saw the turbo core options and I thought that the turbo key mapping be an old codepath not used, but now I see the turbo/R2 key does have a function... I'll fix that conflict and post again

**Comment by markwkidd** [comment]
Thanks again @retro-wertz -- I've pushed a second commit that cleans up the R2/Turbo functionality.

I'm trying to find a good 4+ player game to use as the reference point for this PR when testing joypads.

So far I've tried various NBA Jam and NBA Jam Tournament Edition titles and the two 6-player X-Men titles. None of them seems to be working well enough to serve the purpose. Any suggestions from your experience?

**Comment by wonder199999** [comment]
I also tried to support 4 players  ，：）

https://github.com/wonder199999/mame2010-mini/commit/f49373cd2067b00ef909c468649feafec1c64ce4
https://github.com/wonder199999/mame2010-mini/commit/fef22c82651cc467ec5491716e92182acd628ef8

**Comment by markwkidd** [comment]
Thanks @wonder199999 !!! I can see a few ideas in your code already.

Can I ask did you hit an obstacle in finishing your project? Understandable if it was the time required!

**Comment by markwkidd** [comment]
To answer my own question earlier, `simpsons.zip` seems like a good game to test this with. I've just played a round as Bart (Player 3) with my Buffalo iSNES controller and things seem right!

**Comment by wonder199999** [comment]
@markwkidd
My original intention was to do something similar to 'fbalpha2012-cps''fbalpha2012-neogeo' .
Then add support for more ROMs , as well as a large number of HACK ROMs .
Unfortunately ,it has been extremely short of time recently .

**Comment by markwkidd** [comment]
New clean PR: https://github.com/libretro/mame2010-libretro/pull/103

**Pull request closed without merging**